### PR TITLE
 [FLINK-8856] [TaskManager] Move all cancellation interrupt calls to TaskCanceller thread

### PR DESCRIPTION
**Pull request by StephanEwen**
## What is the purpose of the change

This changes the responsibilities of the threads involved in cancelling a task execution.

The PR cleans up the code and guards against a JVM bug where `interrupt()` calls
block/deadlock if the thread is engaged in certain I/O operations.

We now use three threads potentially in the cancellation
  - *Cancel thread*, calls `cancel()` on the invokable and closes input/output gates. This pretty much as before.
  - The *interrupting thread* calls repeated interrupts after a while, to get the thread out of blocking I/O operations. This helps when the user code is engaged in I/O (directly on through a library) and does not handle interrupts correctly.
  - The *watchdog thread* which kills the TM if cancellation does not succeed after some time (3 minutes default).

Previously, thread two and three were the same, but we saw cases where the watchdog deadlocked (due to a JVM bug) and the hard task cancellation never happened.

Some minor robustness enhancements related to this change are included in this PR.

## Verifying this change

The change is motivated by an occasional JVM bug that I could not purposefully trigger in tests to guard against rollback to the prior state. All tests were passing prior to this change and are passing after this change.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (**yes** / no / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by StephanEwen** [comment]
CI failure unrelated, due to https://issues.apache.org/jira/browse/FLINK-8882

File: flink-core/src/main/java/org/apache/flink/util/ExceptionUtils.java [hunk 1]
```diff
@@ -81,12 +81,15 @@ public static String stringifyException(final Throwable e) {
 	 * <p>Currently considered fatal exceptions are Virtual Machine errors indicating
 	 * that the JVM is corrupted, like {@link InternalError}, {@link UnknownError},
 	 * and {@link java.util.zip.ZipError} (a special case of InternalError).
+	 * The {@link ThreadDeath} exception is also treated as a fatal error, because when
+	 * a thread is forcefully stopped, there is a high chance that parts of the system
+	 * is in an inconsistent state.
```
Comment (aljoscha) [inline flink-core/src/main/java/org/apache/flink/util/ExceptionUtils.java]: nit/typo: "are in an inconsistent state"?

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/DispatcherThreadFactory.java [hunk 2]
```diff
@@ -18,6 +18,8 @@
 
 package org.apache.flink.runtime.taskmanager;
```
Comment (aljoscha) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/DispatcherThreadFactory.java]: nit: the commit message for this one has types: "hanle" and "esure" (feel free to ignore)

**Comment by tillrohrmann** [comment]
Please close this PR @StephanEwen since the commits have been merged.

**Pull request closed without merging**